convert-hf-to-gguf.py

```diff
@@ -409,87 +419,36 @@ def get_vocab_base_pre(self, tokenizer) -> str:
-            # ref: https://huggingface.co/openai-community/gpt2
-            res = "gpt-2"
-        if chkhsh == "6221ad2852e85ce96f791f476e0b390cf9b474c9e3d1362f53a24a06dc8220ff":
-            # ref: https://huggingface.co/smallcloudai/Refact-1_6-base
-            res = "refact"
-        if chkhsh == "9c2227e4dd922002fb81bde4fc02b0483ca4f12911410dee2255e4987644e3f8":
-            # ref: https://huggingface.co/CohereForAI/c4ai-command-r-v01
-            res = "command-r"
-        if chkhsh == "e636dc30a262dcc0d8c323492e32ae2b70728f4df7dfe9737d9f920a282b8aea":
-            # ref: https://huggingface.co/Qwen/Qwen1.5-7B
-            res = "qwen2"
-        if chkhsh == "b6dc8df998e1cfbdc4eac8243701a65afe638679230920b50d6f17d81c098166":
-            # ref: https://huggingface.co/allenai/OLMo-1.7-7B-hf
-            res = "olmo"
-        if chkhsh == "a8594e3edff7c29c003940395316294b2c623e09894deebbc65f33f1515df79e":
-            # ref: https://huggingface.co/databricks/dbrx-base
-            res = "dbrx"
-        if chkhsh == "0876d13b50744004aa9aeae05e7b0647eac9d801b5ba4668afc01e709c15e19f":
-            # ref: https://huggingface.co/jinaai/jina-embeddings-v2-base-en
-            res = "jina-v2-en"
-        if chkhsh == "171aeeedd6fb548d418a7461d053f11b6f1f1fc9b387bd66640d28a4b9f5c643":
-            # ref: https://huggingface.co/jinaai/jina-embeddings-v2-base-es
-            res = "jina-v2-es"
-        if chkhsh == "27949a2493fc4a9f53f5b9b029c82689cfbe5d3a1929bb25e043089e28466de6":
-            # ref: https://huggingface.co/jinaai/jina-embeddings-v2-base-de
-            res = "jina-v2-de"
-
-        if res is None:
-            logger.warning("\n")
-            logger.warning("**************************************************************************************")
-            logger.warning("** WARNING: The BPE pre-tokenizer was not recognized!")
-            logger.warning("**          There are 2 possible reasons for this:")
-            logger.warning("**          - the model has not been added to convert-hf-to-gguf-update.py yet")
-            logger.warning("**          - the pre-tokenization config has changed upstream")
-            logger.warning("**          Check your model files and convert-hf-to-gguf-update.py and update them accordingly.")
-            logger.warning("** ref:     https://github.com/ggerganov/llama.cpp/pull/6920")
-            logger.warning("**")
-            logger.warning(f"** chkhsh:  {chkhsh}")
-            logger.warning("**************************************************************************************")
-            logger.warning("\n")
-            raise NotImplementedError("BPE pre-tokenizer was not recognized - update get_vocab_base_pre()")
-
-        logger.debug(f"tokenizer.ggml.pre: {repr(res)}")
-        logger.debug(f"chkhsh: {chkhsh}")
-
-        return res
+        checksum = sha256(str(tokenizer.vocab).encode()).hexdigest()
```

Comment:
It seems that the checksum will not include the pre-tokenizer config? This does not seem correct because there could be 2 models with the same vocabulary, but different pre-tokenizers - we need to distinguish those

Reply:
If my understanding is correct -correct me if I'm wrong-, the vocab is the set of merges representing the vocabulary. Following this premise would mean the vocab was already normalized and pre-tokenized as these steps are supposed to occur **_before_** training. If you're concerned about the regular expression affecting the encoding and decoding (**_after_** training), I can dig into that and see if that's truly the case. If it is, it would be implementation specific, not necessarily a requirement, e.g. it was a design choice for further processing.

Reply:
Yeah, so after reviewing what I could, the way I'm interpreting is that the pre-tokenizer is the set of regular expressions used to pre-process the text input. We can map this, write it to the model, and then read it as necessary. This allows us to pass around the vocab in the front-end API. The vocab can be hashed as it's already the trained tokenizer. Otherwise, it would need to be derived, e.g. NFD, NFC, etc. aka the normalizers.

Reply:
Sources:
- unicode.org/reports/tr15/
- unicode.org/reports/tr15/#Detecting_Normalization_Forms
- unicode.org/versions/Unicode15.1.0/
  - 3. Conformance
- https://huggingface.co/learn/nlp-course/chapter6/4
- https://huggingface.co/docs/tokenizers/components#pretokenizers
- https://huggingface.co/docs/tokenizers/api/pre-tokenizers

Reply:
@ggerganov 

If we want to create unique identifiers for models based on their vocabularies and pre-tokenizers, a better solution would be to combine both pieces of information into one string before computing the hash. For example:

1. Serialize or encode the tokenizer's vocabulary data
2. Append the specific pre-tokenizer configuration
3. Compute the hash from this concatenated string

By doing so, we ensure that each model with a unique combination of vocabulary and pre-tokenizer configuration will have its distinct identifier in the GGUF file during conversion.

I think I might have a better idea than this though. We can instead hash the pre-tokenizer itself and then save that instead.

I'm currently experimenting with a few approaches. I'm open to ideas for validation.